analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConstCodeFix.cs

```diff
@@ -0,0 +1,73 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[ExportCodeFixProvider(LanguageNames.CSharp)]
+public sealed class UnchangedLocalVariablesShouldBeConstCodeFix : SonarCodeFix
+{
+    private const string Title = "Convert to constant.";
+
+    public override ImmutableArray<string> FixableDiagnosticIds =>
+        ImmutableArray.Create(UnchangedLocalVariablesShouldBeConst.DiagnosticId);
+
+    protected override async Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext context)
+    {
+        if (VariableDeclaration(root, context) is { } variable
+            && variable.Parent is LocalDeclarationStatementSyntax oldNode)
+        {
+            var declaration = variable.Type.IsVar
+                ? WithExplictType(variable, await context.Document.GetSemanticModelAsync().ConfigureAwait(false))
+                : variable;
+            var newNode = root.ReplaceNode(oldNode, ConstantDeclaration(declaration));
```

Comment:
This work should be postponed until it is really needed. It needs to go into the `createChangedDocument` delegate.

Motivation:
`RegisterCodeFixesAsync` is called by the IDE when the QuickFix menu is opened. All that's needed then is:

* Is a quick fix available for this particular diagnostic? If yes register the quick fix
* The title of the quick fix.

Anything that goes beyond that information should only be calculated if the IDE applies the quick fix or displays the preview of the quick fix.

Reply:
It makes sense. I was not aware of this.